lib/spack/spack/util/elf.py

```diff
@@ -0,0 +1,455 @@
+    if sys.version_info[0] < 3:
+        return ord(byte_array[idx])
+    return byte_array[idx]
+
+
+class ElfParsingError(Exception):
+    pass
+
+
+class ElfFile(object):
+    """Parsed ELF file."""
+
+    __slots__ = [
+        "is_64_bit",
+        "is_little_endian",
+        "byte_order",
+        "elf_hdr",
+        "pt_load",
+        # pt_interp
+        "has_pt_interp",
+        "pt_interp_p_offset",
+        "pt_interp_p_filesz",
+        "pt_interp_str",
+        # pt_dynamic
+        "has_pt_dynamic",
+        "pt_dynamic_p_offset",
+        "pt_dynamic_p_filesz",
+        # rpath
+        "has_rpath",
+        "dt_rpath_offset",
+        "dt_rpath_str",
+        "rpath_strtab_offset",
+        "is_runpath",
+        # dt needed
+        "has_needed",
+        "dt_needed_strtab_offsets",
+        "dt_needed_strs",
+        # dt soname
+        "has_soname",
+        "dt_soname_strtab_offset",
+        "dt_soname_str",
+    ]
+
+    def __init__(self):
+        self.dt_needed_strtab_offsets = []
+        self.has_soname = False
+        self.has_rpath = False
+        self.has_needed = False
+        self.pt_load = []
+        self.has_pt_dynamic = False
+        self.has_pt_interp = False
+
+
+def parse_c_string(byte_string, start=0):
+    """
+    Retrieve a C-string at a given offset in a byte string
+
+    Arguments:
+        byte_string (bytes): String
+        start (int): Offset into the string
+
+    Returns:
+        bytes: A copy of the C-string excluding the terminating null byte
+    """
+    str_end = byte_string.find(b"\0", start)
+    if str_end == -1:
+        raise ElfParsingError("C-string is not null terminated")
+    return byte_string[start:str_end]
+
+
+def read_exactly(f, num_bytes, msg):
+    """
+    Read exactly num_bytes at the current offset, otherwise raise
+    a parsing error with the given error message.
+
+    Arguments:
+        f: file handle
+        num_bytes (int): Number of bytes to read
+        msg (str): Error to show when bytes cannot be read
+
+    Returns:
+        bytes: the ``num_bytes`` bytes that were read.
+    """
+    data = f.read(num_bytes)
+    if len(data) != num_bytes:
+        raise ElfParsingError(msg)
+    return data
+
+
+def parse_program_headers(f, elf):
+    """
+    Parse program headers
+
+    Arguments:
+        f: file handle
+        elf (ElfFile): ELF file parser data
+    """
+    # Forward to the program header
+    f.seek(elf.elf_hdr.e_phoff)
+
+    # Here we have to make a mapping from virtual address to offset in the file.
+    program_header_fmt = elf.byte_order + ("LLQQQQQQ" if elf.is_64_bit else "LLLLLLLL")
+    ProgramHeader = ProgramHeader64 if elf.is_64_bit else ProgramHeader32
+    program_header_fmt_size = calcsize(program_header_fmt)
+
+    for _ in range(elf.elf_hdr.e_phnum):
+        data = read_exactly(f, program_header_fmt_size, "Malformed program header")
+        ph = ProgramHeader._make(unpack(program_header_fmt, data))
```

Comment:
Not a big deal, but why read_exactly in a loop rather than something like the following so we can benefit from a larger contiguous read?

```python
full_data = read_exactly(f, phfs * elf.elf_hdr.e_phnum, "badness")
unpack_fron(phf, full_data, i * phfs)
```

Reply:
Sounds good, Python also has `iter_unpack` (but we can only use that after dropping python 2)

Reply:
Right, that would be nice too but we're a bit stuck there for now.